src/cudadev/CUDADataFormats/SiPixelClustersCUDA.h

```diff
@@ -22,41 +32,21 @@ class SiPixelClustersCUDA {
 
   uint32_t nClusters() const { return nClusters_h; }
 
-  uint32_t *moduleStart() { return moduleStart_d.get(); }
-  uint32_t *clusInModule() { return clusInModule_d.get(); }
-  uint32_t *moduleId() { return moduleId_d.get(); }
-  uint32_t *clusModuleStart() { return clusModuleStart_d.get(); }
-
-  uint32_t const *moduleStart() const { return moduleStart_d.get(); }
-  uint32_t const *clusInModule() const { return clusInModule_d.get(); }
-  uint32_t const *moduleId() const { return moduleId_d.get(); }
-  uint32_t const *clusModuleStart() const { return clusModuleStart_d.get(); }
-
-  class DeviceConstView {
-  public:
-    __device__ __forceinline__ uint32_t moduleStart(int i) const { return __ldg(moduleStart_ + i); }
-    __device__ __forceinline__ uint32_t clusInModule(int i) const { return __ldg(clusInModule_ + i); }
-    __device__ __forceinline__ uint32_t moduleId(int i) const { return __ldg(moduleId_ + i); }
-    __device__ __forceinline__ uint32_t clusModuleStart(int i) const { return __ldg(clusModuleStart_ + i); }
+  uint32_t *moduleStart() { return deviceStore_.moduleStart(); }
+  uint32_t *clusInModule() { return deviceStore_.clusInModule(); }
+  uint32_t *moduleId() { return deviceStore_.moduleId(); }
+  uint32_t *clusModuleStart() { return deviceStore_.clusModuleStart(); }
 
-    uint32_t const *moduleStart_;
-    uint32_t const *clusInModule_;
-    uint32_t const *moduleId_;
-    uint32_t const *clusModuleStart_;
-  };
+  uint32_t const *moduleStart() const { return deviceStore_.moduleStart(); }
+  uint32_t const *clusInModule() const { return deviceStore_.clusInModule(); }
+  uint32_t const *moduleId() const { return deviceStore_.moduleId(); }
+  uint32_t const *clusModuleStart() const { return deviceStore_.clusModuleStart(); }
 
-  DeviceConstView *view() const { return view_d.get(); }
+  const DeviceStore store() const { return deviceStore_; }
```

Comment:
Does `DeviceStore` give read-write access to the data? Assuming yes, in CMSSW this function would not be allowed (and thus I'd like to avoid it here too) because of giving mutable access to Event data product from a `const` function.

Reply:
The `DeviceStore` is mutable, but it is `const` correct. There is an `operator[] (size_t) const` returning a different accessor type (`const_element`). Likewise, the full column pointer accessors (`[const] uint32_t * DeviceStore::x() [const]`) exist in 2 variants and are const-correct.

We can also make a stripped-down version of the store but the compiler should achieve the same with `const`correctness. The framework could even enforce returning const variants with `std::is_const`.

Should we still define a stripped down `DeviceConstStore` ? (I did if for views, but it's redundant to `const` views).

Reply:
I'm afraid `DeviceStore` being `const`-correct is not sufficient, because the following is possible
```cpp
SiPixelClustersCUDA const& clusters = ctx.get(iEvent, ...);
DeviceStore mutableStore = clusters.store();
// 'mutableStore' "leaks" mutable access to the memory buffer owned by `clusters`
```

Reply:
> So we do need the const-only views, and products should only provide those.

Let me just add that for a product class it is ok for a non-`const` function to return a mutable view (or pointer-to-mutable), as long as the product itself is not copyable (like here), or a copy constructor/assignment makes a deep copy.

Reply:
To be complete, we have both const and non-const views. Which one is provided to the user of the product will be up to the product implementation.